### DISCUSSION TITLE
DB の方言に依存せず Sequence を実行する機能について

### DISCUSSION THREAD

**Discussion by ironclad-15256304** [domaframework/doma#1000] [Ideas]
### 相談内容

いつも大変お世話になっております。

私の参加しているプロジェクトでは、 SQL の Sequence 機能を利用し、正確に被りのない連番を必要とする場面がしばしば登場します。
そのようなシチュエーションのために、 Entity のフィールドに `@GeneratedValue` アノテーションを付与することで Insert 実行時に自動で Sequence から ID を作成する機能があるのは承知しております。しかし、データの Insert のタイミングと関係なくシンプルに Sequence のみを実行したい場合があります（設計を見直すべきだとは思いますが、他の方法がないという仮定で進めさせてください）。
そのような場合に、 SQL Template を使用して Sequence を実行してもよいのですが、そうすると `@GeneratedValue(strategy = GenerationType.SEQUENCE)` で吸収されているであろう DB の方言等を考慮しなければなりません。それも大した手間ではないとは認識していますが、可能であればそのあたりを意識しないでいいように、 Doma の力をお借りしたいです。

上記のような状況において、任意の名称の Sequence を実行し、その値を取得するだけの機能を追加していただくことは可能でしょうか。もしくは、既に何かしらの形で実現が可能なのでしょうか。
お忙しいところ大変恐縮ですが、ご検討頂けますと幸いです。

**Comment by nakamura-to** [comment]
以下のDialectのメソッドでDomaが組み立てているシーケンスのSQLを取得できます。

https://github.com/domaframework/doma/blob/2.54.1/doma-core/src/main/java/org/seasar/doma/jdbc/dialect/Dialect.java#L224

このSQLを使ってアプリケーションからSQLを発行してはどうでしょうか？

**Reply by ironclad-15256304** [reply]
なるほど、 Dao 中であれば `Config.get(this).getDialect().getSequenceNextValSql("sequence_name", 1)` のような形で `Sql` が取得できるのですね。
[org.seasar.doma.jdbc.id.AbstractIdGenerator#getGeneratedValue](https://github.com/domaframework/doma/blob/master/doma-core/src/main/java/org/seasar/doma/jdbc/id/AbstractIdGenerator.java#L24) を参考に実装してみようと思ったのですが、 [org.seasar.doma.internal.jdbc.util.JdbcUtil](https://github.com/domaframework/doma/blob/master/doma-core/src/main/java/org/seasar/doma/internal/jdbc/util/JdbcUtil.java) を使用しているようです。浅学で大変恐縮なのですが、こちらを普通に Dao から呼び出して同様の実装を行っても問題ないのでしょうか？

**Comment by nakamura-to** [comment]
`JdbcUtil`はinternalなコードなので使わない方がいいと思います。
こんな感じのコードを書くのはどうでしょうか？

```java
  default int nextValue() {
    Config config = Config.get(this);
    org.seasar.doma.jdbc.Sql<?> sql = config.getDialect().getSequenceNextValSql("MY_SEQUENCE", 0);
    try (Connection c = config.getDataSource().getConnection();
         Statement s = c.createStatement();
         ResultSet r = s.executeQuery(sql.getRawSql())) {
      if (r.next()) {
        return r.getInt(1);
      }
      throw new RuntimeException("no value");
    } catch (SQLException e) {
      throw new RuntimeException(e);
    }
  }
```

**Reply by ironclad-15256304** [reply]
返信が遅れ申し訳ございません。
なるほど、Sql#getRawSql で取れるSQL文を、 java.sql パッケージの部品を使用してそのまま実行してしまえばよいのですね。
internal な部品についてはこちらでは参照せず、ご提示いただいたコードを参考にして実装に組み込んでみようと思います。有難うございました！
